server/src/test/java/org/elasticsearch/plugins/PluginsServiceTests.java

```diff
@@ -613,6 +621,76 @@ public void testThrowingConstructor() {
         assertThat(e.getCause().getCause(), hasToString(containsString("test constructor failure")));
     }
 
+    public void testLoadServiceProviders() throws ClassNotFoundException {
+        ClassLoader fakeClassLoader = mock(ClassLoader.class);
+        ClassLoader fakeClassLoader1 = mock(ClassLoader.class);
+        FakePlugin fakePlugin = new FakePlugin();
+        FakePlugin fakePlugin1 = new FakePlugin();
+
+        OperatorHandlerProvider provider1 = () -> List.of(new OperatorHandler<Integer>() {
+            @Override
+            public String name() {
+                return "integer";
+            }
+
+            @Override
+            public TransformState transform(Object source, TransformState prevState) {
+                return prevState;
+            }
+        });
+
+        OperatorHandlerProvider provider2 = () -> List.of(new OperatorHandler<String>() {
+            @Override
+            public String name() {
+                return "string";
+            }
+
+            @Override
+            public TransformState transform(Object source, TransformState prevState) {
+                return prevState;
+            }
+        });
+        // Make the service load without plugins, we mock and force set them after
+        PluginsService service = newMockPluginsService(new ArrayList<>());
+        PluginsService.LoadedPlugin plugin1 = new PluginsService.LoadedPlugin(
+            mock(PluginDescriptor.class),
+            fakePlugin,
+            fakeClassLoader,
+            ModuleLayer.boot()
+        );
+        PluginsService.LoadedPlugin plugin2 = new PluginsService.LoadedPlugin(
+            mock(PluginDescriptor.class),
+            fakePlugin1,
+            fakeClassLoader1,
+            ModuleLayer.boot()
+        );
+        ((MockPluginsService) service).setLoadedPlugins(List.of(plugin1, plugin2));
+
+        service = spy(service);
+
+        doReturn(List.of(provider1).iterator()).when(service).providersIterator(eq(OperatorHandlerProvider.class), eq(fakeClassLoader));
```

Comment:
Since we are dealing with different classloaders, we should probably test more real-world situations. Can the test use real classloaders instead of mocking them? We have utilities for creating dummy jars with module infos, classes, etc. It seems doable to actually test the service loading.